content/en/platform/corda/5.0-dev-preview-1/cordapps/ledger-dsl.md

```diff
@@ -0,0 +1,342 @@
+
+**Kotlin**
+
+```kotlin
+input(MyState(UniqueIdentifier(), listOf(alice), "input 1"), notary, txId)
+```
+
+**Java**
+
+```java
+dsl.input(new MyState(new UniqueIdentifier(), List.of(alice), "input 1"), notary, txId);
+```
+
+* Adds the passed in `StateAndRef` to the transaction. This `StateAndRef` could come from another transaction or built manually. This method matches `TransactionBuilder.addInputState` and allows you to write code that closer matches your production code.
+
+**Kotlin**
+
+```kotlin
+input(stateAndRef)
+```
+
+**Java**
+
+```java
+dsl.input(stateAndRef);
+```
+
+### `reference`
+
+There are 3 different flavors of this method, explained alongside the snippets below:
+
+* Creates a new `StateAndRef` with a random transaction id and adds it to the transaction. This is the easiest way to add a new reference for testing.
+
+**Kotlin**
+
+```kotlin
+reference(MyState(UniqueIdentifier(), listOf(alice), "input 1"), notary)
+```
+
+**Java**
+
+```java
+dsl.reference(new MyState(new UniqueIdentifier(), List.of(alice), "input 1"), notary);
+```
+
+* Creates a new `StateAndRef` with the passed in transaction id (`txId`) and adds it to the transaction. This is useful if your contract verification validates any transaction ids.
+
+**Kotlin**
+
+```kotlin
+reference(MyState(UniqueIdentifier(), listOf(alice), "input 1"), notary, txId)
+```
+
+**Java**
+
+```java
+dsl.reference(new MyState(new UniqueIdentifier(), List.of(alice), "input 1"), notary, txId);
+```
+
+* Adds the passed in `StateAndRef` to the transaction. This `StateAndRef` could come from another transaction or built manually. This method matches `TransactionBuilder.addReferenceState` and allows you to write code that closer matches your production code.
+
+**Kotlin**
+
+```kotlin
+reference(stateAndRef)
+```
+
+**Java**
+
+```java
+dsl.reference(stateAndRef);
+```
+
+### `attachment`
+
+There are 2 overloads of this method, explained alongside the snippets below:
+
+* Creates an attachment that reads the contents of a file found in the `resource` directory when accessed. This is the simplest version to use as you need provide the name of the file and nothing else.
+
+**Kotlin**
+
+```kotlin
+attachment("important-data.csv")
+```
+
+**Java**
+
+```java
+dsl.attachment("important-data.csv");
+```
+
+*Accepts a mock `Attachment` instance that you can configure as desired.
+
+**Kotlin**
+
+```kotlin
+val mockAttachment: Attachment = mock()
+// mocking of attachment methods
+attachment(mockAttachment)
+```
+
+**Java**
+
+```java
+Attachment mockAttachment = mock(Attachment.class);
+// mocking of attachment methods
+attachment(mockAttachment)
+```
+
+{{< note >}}
+`Attachment.open` returns an `InputStream` representing a ZIP file and you should mock its functionality as such or it will differ from when you run your contracts in production.
```

Comment:
```suggestion
`Attachment.open` returns an `InputStream` representing a ZIP file. You should mock its functionality as such or it will differ from when you run your contracts in production.
```
Not sure about the as such here.